File: Java/ToyORB/src/ToyORB/RemoteObjectReference.java
package ToyORB;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class RemoteObjectReference {
	public final String host;
	public final int port;
	public final String serviceName;
	public final String typeName;

	public RemoteObjectReference(String host, int port, String serviceName, String typeName) {
		this.host = host;
		this.port = port;
		this.serviceName = serviceName;
		this.typeName = typeName;
	}

	public void writeInto(DataOutputStream stream) throws IOException {
		NetUtils.writeStringInto(stream, host);
		stream.writeInt(port);
		NetUtils.writeStringInto(stream, serviceName);
		NetUtils.writeStringInto(stream, typeName);
	}

	public static RemoteObjectReference readFrom(DataInputStream stream) throws IOException {
		String host = NetUtils.readStringFrom(stream);
		int port = stream.readInt();
		String name = NetUtils.readStringFrom(stream);
		String typeName = NetUtils.readStringFrom(stream);
		return new RemoteObjectReference(host, port, name, typeName);
	}

	@Override
	public String toString() {
		return "RemoteObjectReference{" +
				"host='" + host + '\'' +
				", port=" + port +
				", serviceName='" + serviceName + '\'' +
				", typeName='" + typeName + '\'' +
				'}';
	}
}

File: .NET Core/ToyORB/RemoteObjectReference.cs
using System;
using System.IO;
using ToyORB.Messages;

namespace ToyORB
{
    public class RemoteObjectReference
    {
        public string Host { get; }
        public int Port { get; }
        public string ServiceName { get; }
        public string TypeName { get; }

        public RemoteObjectReference(string host, int port, string serviceName, string typeName)
        {
            Host = host ?? throw new ArgumentNullException(nameof(host));
            Port = port;
            ServiceName = serviceName ?? throw new ArgumentNullException(nameof(serviceName));
            TypeName = typeName ?? throw new ArgumentNullException(nameof(typeName));
        }

        public void WriteInto(BinaryWriter stream)
        {
            stream.WriteBigEndian(Host);
            stream.WriteBigEndian(Port);
            stream.WriteBigEndian(ServiceName);
            stream.WriteBigEndian(TypeName);
        }

        public static RemoteObjectReference ReadFrom(BinaryReader stream)
        {
            string host = stream.ReadStringBigEndian();
            int port = stream.ReadInt32BigEndian();
            string serviceName = stream.ReadStringBigEndian();
            string typeName = stream.ReadStringBigEndian();
            return new RemoteObjectReference(host, port, serviceName, typeName);
        }

        public override string ToString()
        {
            return $"RemoteObjectReference{{{nameof(Host)}: {Host}, {nameof(Port)}: {Port}, {nameof(ServiceName)}: {ServiceName}, {nameof(TypeName)}: {TypeName}}}";
        }
    }
}
